tests/pytests/unit/fileclient/test_fileclient_cache.py

```diff
@@ -0,0 +1,342 @@
+import errno
+import logging
+import os
+import shutil
+
+import pytest
+
+import salt.utils.files
+from salt import fileclient
+from tests.support.mock import patch
+
+log = logging.getLogger(__name__)
+
+
+SUBDIR = "subdir"
+
+
+def _saltenvs():
+    return ("base", "dev")
+
+
+def _subdir_files():
+    return ("foo.txt", "bar.txt", "baz.txt")
+
+
+def _get_file_roots(fs_root):
+    return {x: [os.path.join(fs_root, x)] for x in _saltenvs()}
+
+
+@pytest.fixture
+def fs_root(tmp_path):
+    return os.path.join(tmp_path, "fileclient_fs_root")
+
+
+@pytest.fixture
+def cache_root(tmp_path):
+    return os.path.join(tmp_path, "fileclient_cache_root")
+
+
+@pytest.fixture
+def mocked_opts(tmp_path, fs_root, cache_root):
+    return {
+        "file_roots": _get_file_roots(fs_root),
+        "fileserver_backend": ["roots"],
+        "cachedir": cache_root,
+        "file_client": "local",
+    }
+
+
+@pytest.fixture
+def configure_loader_modules(tmp_path, mocked_opts):
+    return {fileclient: {"__opts__": mocked_opts}}
+
+
+@pytest.fixture(autouse=True)
+def setup(fs_root, cache_root):
+    """
+    No need to add a dummy foo.txt to muddy up the github repo, just make
+    our own fileserver root on-the-fly.
+    """
+
+    def _new_dir(path):
+        """
+        Add a new dir at ``path`` using os.makedirs. If the directory
+        already exists, remove it recursively and then try to create it
+        again.
+        """
+        try:
+            os.makedirs(path)
+        except OSError as exc:
+            if exc.errno == errno.EEXIST:
+                # Just in case a previous test was interrupted, remove the
+                # directory and try adding it again.
+                shutil.rmtree(path)
+                os.makedirs(path)
+            else:
+                raise
+
+    # Crete the FS_ROOT
+    for saltenv in _saltenvs():
+        saltenv_root = os.path.join(fs_root, saltenv)
+        # Make sure we have a fresh root dir for this saltenv
+        _new_dir(saltenv_root)
+
+        path = os.path.join(saltenv_root, "foo.txt")
+        with salt.utils.files.fopen(path, "w") as fp_:
+            fp_.write("This is a test file in the '{}' saltenv.\n".format(saltenv))
+
+        subdir_abspath = os.path.join(saltenv_root, SUBDIR)
+        os.makedirs(subdir_abspath)
+        for subdir_file in _subdir_files():
+            path = os.path.join(subdir_abspath, subdir_file)
+            with salt.utils.files.fopen(path, "w") as fp_:
+                fp_.write(
+                    "This is file '{}' in subdir '{} from saltenv '{}'".format(
+                        subdir_file, SUBDIR, saltenv
+                    )
+                )
+
+    # Create the CACHE_ROOT
+    _new_dir(cache_root)
+
+
+def test_cache_dir(mocked_opts, minion_opts):
+    """
+    Ensure entire directory is cached to correct location
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
+    patched_opts.update(mocked_opts)
```

Comment:
```suggestion
    patched_opts = minion_opts.copy()
    patched_opts.update(mocked_opts)
```
In fact, the `minion_opts` fixture is generated(or copied) for every test function, so, you might not even need to copy it...